Sources/MongoSwift/MongoCollection+FindAndModify.swift

```diff
@@ -50,32 +56,61 @@ extension MongoCollection {
      *
      * - Returns: A `CollectionType` representing either the original or updated document,
      *      depending on selected options, or `nil` if there was no match.
-     * - Throws: 
+     *
+     * - Throws:
+     *   - `MongoError.invalidArgument` if any of the provided options are invalid
      *   - `MongoError.commandError` if there are any errors executing the command.
      *   - A `DecodingError` if the updated document cannot be decoded to a `CollectionType` value
      */
     @discardableResult
     public func findOneAndUpdate(filter: Document, update: Document,
                                  options: FindOneAndUpdateOptions? = nil) throws -> CollectionType? {
-        throw MongoError.commandError(message: "Unimplemented command")
+        return try self.findAndModify(filter: filter, update: update, options: options)
+    }
+
+    /// A private helper method for findAndModify operations to use
+    private func findAndModify(filter: Document, update: Document? = nil,
+                               options: FindAndModifyOptionsConvertible? = nil) throws -> CollectionType? {
+
+        var opts = try options?.asOpts()
+
+        if let update = update {
+            // if no options were provided but we need to set update, create them
+            opts = opts ?? FindAndModifyOptions()
+            try opts!.setUpdate(update)
+        }
+
+        let reply = Document()
+        var error = bson_error_t()
+
+        if !mongoc_collection_find_and_modify_with_opts(self._collection,
+        filter.data, opts?._options, reply.data, &error) {
+            // TODO SWIFT-144: replace with more descriptive error type(s)
+            throw MongoError.commandError(message: toErrorString(error))
+        }
+
+        guard let value = reply["value"] as? Document else { return nil }
+
+        return try BsonDecoder().decode(CollectionType.self, from: value)
     }
 }
 
 /// Indicates which document to return in a find and modify operation.
-public enum ReturnDocument: Encodable {
+public enum ReturnDocument {
     /// Indicates to return the document before the update, replacement, or insert occured.
     case before
-
     ///  Indicates to return the document after the update, replacement, or insert occured.
     case after
+}
 
-    public func encode(to encoder: Encoder) throws {
-        // fill in later on
-    }
+/// Indicates that an options type can be represented as a `FindAndModifyOptions`
+private protocol FindAndModifyOptionsConvertible {
```

Comment:
What does a private protocol entail? Wouldn't `internal` make more sense here to limit it to structs/classes within this file? While implementing the bulk write models, I was experimenting with:

```swift
internal protocol WriteModel {
    func addToBulkWrite(bulk: OpaquePointer)
}
```

As I understand it, `internal` on the protocol will require the `addToBulkWrite()` method to be `internal` on implementing structs.

Reply:
top-level `private` declarations are accessible throughout the file they're defined in -- basically, top level `private` = `fileprivate`. 
all the public types conforming to the protocol are in this file, so they can access the protocol. but they need to explicitly mark their `asOpts` implementations as `fileprivate` to match the protocol. 

in contrast, making a protocol `internal` would allow it to be used in any of our source files.